### ISSUE TITLE
Discussion: Change default divulgence behaviour of choices

### ISSUE THREAD

**Issue by bame-da** [digital-asset/daml#1300]
Consider this scenario:
```
template Secret
  with
    p : Party
  where
    signatory p

template SecretFetcher
  with
    p : Party
    obs : [Party]
  where 
    signatory p
    observer obs

    controller p can
      PublicFetch : Secret
        with s : ContractId Secret
        do fetch s

      preconsuming PreconsumingFetch : Secret
        with s : ContractId Secret
        do fetch s

      postconsuming PostconsumingFetch : Secret
        with s : ContractId Secret
        do fetch s

      nonconsuming CustomPreconsumingFetch : Secret
        with s : ContractId Secret
        do 
          archive self
          fetch s

      nonconsuming CustomPostconsumingFetch : Secret
        with s : ContractId Secret
        do 
          x <- fetch s
          archive self
          return x

privacytest = scenario do
  p <- getParty "p"
  o <- getParty "o"

  submit p do
    s <- create Secret with p
    sf <- create SecretFetcher with p; obs = [o]
    exercise sf PublicFetch with s

    s <- create Secret with p
    sf <- create SecretFetcher with p; obs = [o]
    exercise sf PreconsumingFetch with s

    s <- create Secret with p
    sf <- create SecretFetcher with p; obs = [o]
    exercise sf PostconsumingFetch with s

    s <- create Secret with p
    sf <- create SecretFetcher with p; obs = [o]
    exercise sf CustomPreconsumingFetch with s

    s <- create Secret with p
    sf <- create SecretFetcher with p; obs = [o]
    exercise sf CustomPostconsumingFetch with s```
```
This results in the first two secrets to be shown to `o` and the last three not. This is _very_ confusing. I would propose that we compile normal consuming and `preconsuming` choices to a nonconsuming choice with `archive self` at the top.

It _may_ sometimes be desireable to show observers exercises. Maybe we should introduce a new keyword `observable` to mark choices as being observable to non-authorizers, or we should add an explicit disclosure mechanism so that I can explicitly show sub-transactions to parties (e.g. `disclose : [Party] -> Update a -> Update a

**Comment by bame-da** [comment]
The main problem is that this may well break workflows that rely on divulgence. This may deserve a version bump to 1.3 if we do it.

**Comment by oggy-** [comment]
I agree that the current privacy model is not obvious, especially with pre- and post-consuming choices and I'm happy to see it improved. I leave it up to you guys to come up with the best constructs for doing so.

However, the proposed mechanism further divorces the surface language from the core language and the ledger model. So far, we already have the following non-obvious steps in the translation of the surface language:
- every exercise desugars into a fetch + exercise (though this might go away according to @hurryabit )
- an archive desugars into a built-in choice with no consequences
- a post-consuming exercise desugars into a non-consuming exercise + archive

Keeping the ledger model minimal might make it easier to implement, but I think it makes it less useful as a mental model for the users of the language. This is something that we need to make a conscious decision on.

**Comment by stefanobaghino-da** [comment]
> This results in the first two secrets to be shown to o and the last three not. This is very confusing. I would propose that we compile normal consuming and preconsuming choices to a nonconsuming choice with archive self at the top.

With regards to this part of the discussion, I believe #4164 (documented in #4134) addresses it in full.
